examples/cfd/swe_nonlinear_pino/train_swe_nl_pino.py

```diff
@@ -0,0 +1,263 @@
+        # wrap epoch in launch logger for console logs
+        with LaunchLogger(
+            "train",
+            epoch=epoch,
+            num_mini_batch=len(train_loader),
+            epoch_alert_freq=10,
+        ) as log:
+            for x, y in train_loader:
+                # Pad input to be put into model
+                x, y = x.to("cuda"), y.to("cuda")
+                x_in = F.pad(x, (0, 0, 0, padding), "constant", 0)
+                x_in = x_in.permute(0, 4, 3, 1, 2)
+                # compute forward pass and unpad output
+                out = (
+                    model(x_in)
+                    .permute(0, 3, 4, 2, 1)
+                    .reshape(batch_size, S, S, T + padding, nfields)
+                )
+                out = out[..., :-padding, :]
+                s0 = x[..., 0, -1]
+
+                # Compute PDE loss using 'modulus' functions or method from 'original' paper
+                if cfg.loss.derivative == "original":
+                    loss_pde = PINO_loss_swe_nonlin(
+                        out,
+                        g=cfg.data.g,
+                        nu=cfg.data.nu,
+                        h_weight=cfg.loss.h_loss,
+                        u_weight=cfg.loss.u_loss,
+                        v_weight=cfg.loss.v_loss,
+                    )
+                elif cfg.loss.derivative == "modulus":
+                    loss_pde = modulus_fourier(
+                        out,
+                        swe_nl_node,
+                        h_weight=cfg.loss.h_loss,
+                        u_weight=cfg.loss.u_loss,
+                        v_weight=cfg.loss.v_loss,
+                    )
+
+                # Compute data loss
+                loss_l2 = swe_loss(out, y, H=cfg.data.H)
+
+                # Compute initial condition loss
+                loss_ic = ic_loss(out, s0)
+
+                # Compute total loss
+                loss = (
+                    cfg.loss.ic_loss * loss_ic
+                    + cfg.loss.xy_loss * loss_l2
+                    + cfg.loss.f_loss * loss_pde
+                )
+
+                # Backward pass and optimizer
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+                log.log_minibatch(
+                    {
+                        "train loss": loss,
+                        "train loss_ic": loss_ic.detach(),
+                        "train loss_l2": loss_l2.detach(),
+                        "train loss_pde": loss_pde.detach(),
+                    }
+                )
+            # Learning rate update
+            scheduler.step()
+            log.log_epoch({"Learning Rate": optimizer.param_groups[0]["lr"]})
+        save_checkpoint(
+            "./checkpoints",
+            models=model,
+            optimizer=optimizer,
+            scheduler=scheduler,
+            epoch=epoch,
+        )
+
+    # Test model
+    with LaunchLogger("test", epoch=epoch) as log:
```

Comment:
Just wondering if we can call this after a few x epochs during the training? Since this step is also visualizing the results, would be nice to view the outputs as the training progresses.

Reply:
changed to call it every 15 epochs and included a generate_movie function and the end to generate a .gif of the predictions and error

Reply:
Thanks for adding this!